Add API stubs for sidePanel and sidebarAction.

AriYoung00 opened the pull request:
#### 85f7152d1097d85d636268bbf0ab96126fc9e610
<pre>
Add API stubs for sidePanel and sidebarAction.
<a href="https://webkit.org/b/276799">https://webkit.org/b/276799</a>
<a href="https://rdar.apple.com/131488956">rdar://131488956</a>

Reviewed by Timothy Hatcher.

This patch adds IDL files defining the sidebarAction and sidePanel JS
extension APIs. It also changes some project settings and infrastructure
to include these files in the build process. Finally, it also adds them
to the `browser` namespace in WebExtensionAPINamespace.idl and adds
additional machinery and stubs in related code to ensure everything
functions as intended. I tested this patch by compiling with
WK_WEB_EXTENSIONS_SIDEBAR enabled (and the properties enabled in
isPropertyAllowed) and verifying that the behavior from JS in a web
extension context is as intended (namely, returning an &quot;unimplemented&quot;
error).

* Source/WebKit/DerivedSources-input.xcfilelist: Add
  WebExtensionAPISidebarAction.idl and WebExtensionAPISidePanel.idl
* Source/WebKit/DerivedSources-output.xcfilelist: Add
  JSWebExtensionAPISidebarAction.{h,mm} and
  JSWebExtensionAPISidePanel.{h,mm}
* Source/WebKit/DerivedSources.make: Add WebExtensionAPISidebarAction
  and WebExtensionAPISidePanel interfaces
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Update project to
  include new files and derived sources
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Add cases for
sidebarAction and sidePanel properties (stubbed, both return `false` for
now).
(WebKit::WebExtensionAPINamespace::sidebarAction): Add getter for
m_sidebarAction
(WebKit::WebExtensionAPINamespace::sidePanel): Add getter for
m_sidePanel
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm: Added.
(WebKit::WebExtensionAPISidePanel::getOptions): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidePanel::setOptions): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidePanel::getPanelBehavior): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidePanel::setPanelBehavior): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidePanel::open): Add stub for API
function returning &apos;Error: Unimplemented&apos;
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm: Added.
(WebKit::WebExtensionAPISidebarAction::open): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::isOpen): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::close): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::toggle): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::getPanel): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::setPanel): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::getTitle): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::setTitle): Add stub for API
function returning &apos;Error: Unimplemented&apos;
(WebKit::WebExtensionAPISidebarAction::setIcon): Add stub for API
function returning &apos;Error: Unimplemented&apos;
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
  Add m_sidebarAction and m_sidePanel private members, add declaration
  for corresponding getters.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidePanel.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Add namespace properties for sidebarAction and sidePanel.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidePanel.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidebarAction.idl: Added.

Canonical link: <a href="https://commits.webkit.org/281149@main">https://commits.webkit.org/281149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de510a12471e70b7d2ebe785c0dd21df2df19dab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58939 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38268 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62572 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9383 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45919 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9585 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/6681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60970 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/35797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/50971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28520 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/32524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8387 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/54478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8542 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64272 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2852 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54985 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2862 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50998 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/55087 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2390 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8798 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/34097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/36266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/34927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/DerivedSources.make:
```diff
@@ -885,6 +885,8 @@ EXTENSION_INTERFACES = \
     WebExtensionAPIPort \
     WebExtensionAPIRuntime \
     WebExtensionAPIScripting \
+    WebExtensionAPISidebarAction \
```
nit: `WebExtensionAPISidePanel \` should technically come first since Side < Sidebar. the way i always double check ordering is copying the lines in question and running `pbpaste | sort | pbcopy` in terminal!

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -74,6 +74,16 @@
         return !extensionContext().supportsManifestVersion(3) && objectForKey<NSDictionary>(extensionContext().manifest(), @"page_action", false);
 
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+    // TODO: check if sidebar feature is enabled and if we have a sidebarAction entry in the manifest to determine if we can use sidebarAction
```
nit: we usually use `// FIXME: rdar://<123456>  Check if sidebar feature is enabled ...`

kiaraarose replied:
or if this is being tracked [www.bugs.webkit.org](https://bugs.webkit.org), then the link to the webkit bug

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
```diff
@@ -220,6 +230,28 @@
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+WebExtensionAPISidebarAction& WebExtensionAPINamespace::sidebarAction()
+{
+    // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/sidebarAction
```
nit: can drop the en-US
// Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/sidebarAction

kiaraarose commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidePanel.idl:
```diff
@@ -0,0 +1,40 @@
+ */
+
+[
+    Conditional=WK_WEB_EXTENSIONS & WK_WEB_EXTENSIONS_SIDEBAR,
```
i see examples with no space in the codebase. does it matter? i.e. `WK_WEB_EXTENSIONS&WK_WEB_EXTENSIONS_SIDEBAR`

AriYoung00 replied:
I don't believe so, the generated files seem to have the correct guard:
```
#if ENABLE(WK_WEB_EXTENSIONS) && ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
...
```

xeenon replied:
You can just use the sidebar define though, since it depends on the other.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
```diff
@@ -60,6 +60,11 @@
 
     [MainWorldOnly, Dynamic] readonly attribute WebExtensionAPIScripting scripting;
 
+    // TODO: Figure out if we want to try to figure out how to use EnabledBySetting=WebExtensionSidebarEnabled for this and the next
```
I forgot to remove this, please ignore the errant TODO

kiaraarose commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidePanel.idl:
```diff
@@ -0,0 +1,40 @@
+    ReturnsPromiseWhenCallbackIsOmitted,
+] interface WebExtensionAPISidePanel {
+
+    [RaisesException] void getOptions([Optional, NSDictionary] any options, [Optional, CallbackHandler] function callback);
```
don't think the options is `optional` based on the chrome spec

AriYoung00 replied:
Yep you're right, I'll go back and remove these `optional`s

kiaraarose commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidePanel.idl:
```diff
@@ -0,0 +1,40 @@
+    [RaisesException] void getPanelBehavior([Optional, CallbackHandler] function callback);
+    [RaisesException] void setPanelBehavior([NSDictionary] any behavior, [Optional, CallbackHandler] function callback);
+
+    [RaisesException] void open([Optional, NSDictionary] any options, [Optional, CallbackHandler] function callback);
```
same here

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -0,0 +1,67 @@
+#import "config.h"
+#import "WebExtensionAPISidePanel.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS) && ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
```
Just doing the sidebar feature is enough, since it is defined as requiring the other already. Same goes for the other spots in this PR.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
2024

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
```diff
@@ -0,0 +1,86 @@
+
+#if ENABLE(WK_WEB_EXTENSIONS) && ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+
+#import "Logging.h"
```
Is this needed?

AriYoung00 replied:
Whoops, no -- forgot to remove all of these `Logging.h` imports when I removed my debugging log statements. Removing now!

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm:
```diff
@@ -0,0 +1,67 @@
+
+#if ENABLE(WK_WEB_EXTENSIONS) && ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+
+#import "Logging.h"
```
Is this needed?

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidebarAction.idl:
```diff
@@ -0,0 +1,43 @@
+ */
+
+[
+    Conditional=WK_WEB_EXTENSIONS & WK_WEB_EXTENSIONS_SIDEBAR,
```
Just the sidebar define.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidebarAction.idl:
```diff
@@ -0,0 +1,43 @@
+    [RaisesException] void close([Optional, CallbackHandler] function callback);
+    [RaisesException] void toggle([Optional, CallbackHandler] function callback);
+
+    [RaisesException] void getPanel([NSDictionary=NullAllowed] any details, [Optional, CallbackHandler] function callback);
```
does `[NSDictionary=NullAllowed]` mean the object passed is allowed to be null?

AriYoung00 replied:
I believe that it means the values within it are allowed to be null, e.g. from JS we could pass in a `details` object like `{ tabId: null }`.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
```diff
@@ -79,6 +81,10 @@ class WebExtensionAPINamespace : public WebExtensionAPIObject, public JSWebExten
     WebExtensionAPIRuntime& runtime() const final;
     WebExtensionAPIScripting& scripting();
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+    WebExtensionAPISidebarAction& sidebarAction();
```
nit: sorting

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
```diff
@@ -103,6 +109,10 @@ class WebExtensionAPINamespace : public WebExtensionAPIObject, public JSWebExten
     mutable RefPtr<WebExtensionAPIRuntime> m_runtime;
     RefPtr<WebExtensionAPIScripting> m_scripting;
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+    RefPtr<WebExtensionAPISidebarAction> m_sidebarAction;
```
ditto

AriYoung00 committed:
Add API stubs for sidePanel and sidebarAction.
https://webkit.org/b/276799
rdar://131488956

Reviewed by Timothy Hatcher.

This patch adds IDL files defining the sidebarAction and sidePanel JS
extension APIs. It also changes some project settings and infrastructure
to include these files in the build process. Finally, it also adds them
to the `browser` namespace in WebExtensionAPINamespace.idl and adds
additional machinery and stubs in related code to ensure everything
functions as intended. I tested this patch by compiling with
WK_WEB_EXTENSIONS_SIDEBAR enabled (and the properties enabled in
isPropertyAllowed) and verifying that the behavior from JS in a web
extension context is as intended (namely, returning an "unimplemented"
error).

* Source/WebKit/DerivedSources-input.xcfilelist: Add
  WebExtensionAPISidebarAction.idl and WebExtensionAPISidePanel.idl
* Source/WebKit/DerivedSources-output.xcfilelist: Add
  JSWebExtensionAPISidebarAction.{h,mm} and
  JSWebExtensionAPISidePanel.{h,mm}
* Source/WebKit/DerivedSources.make: Add WebExtensionAPISidebarAction
  and WebExtensionAPISidePanel interfaces
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Update project to
  include new files and derived sources
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed): Add cases for
sidebarAction and sidePanel properties (stubbed, both return `false` for
now).
(WebKit::WebExtensionAPINamespace::sidebarAction): Add getter for
m_sidebarAction
(WebKit::WebExtensionAPINamespace::sidePanel): Add getter for
m_sidePanel
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidePanelCocoa.mm: Added.
(WebKit::WebExtensionAPISidePanel::getOptions): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidePanel::setOptions): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidePanel::getPanelBehavior): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidePanel::setPanelBehavior): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidePanel::open): Add stub for API
function returning 'Error: Unimplemented'
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm: Added.
(WebKit::WebExtensionAPISidebarAction::open): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::isOpen): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::close): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::toggle): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::getPanel): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::setPanel): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::getTitle): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::setTitle): Add stub for API
function returning 'Error: Unimplemented'
(WebKit::WebExtensionAPISidebarAction::setIcon): Add stub for API
function returning 'Error: Unimplemented'
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
  Add m_sidebarAction and m_sidePanel private members, add declaration
  for corresponding getters.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidePanel.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Add namespace properties for sidebarAction and sidePanel.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidePanel.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidebarAction.idl: Added.

Canonical link: https://commits.webkit.org/281149@main
